### PR TITLE
Introduce Symfony UX Turbo

### PR DESCRIPTION
**Pull request by tgalopin**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Tickets       | -
| License       | MIT

This PR introduces a new package to Symfony UX: UX Turbo, integrating the [Hotwire Turbo](https://turbo.hotwire.dev) library with Symfony. It is based on an initial prototype made by Kevin Dunglas (@dunglas) as a [Mercure](https://mercure.rocks/) project.

Hotwire is an aggregation of several technologies developed by the Ruby on Rails ecosystem and the Bascamp company to build interactive applications with as less JavaScript code as possible. Stimulus, as the foundation of Symfony UX, is already well integrated into Symfony. This PR goes to the next step: integrating Turbo.

You can learn more about the Hotwire intiative on https://twitter.com/dhh/status/1341420143239450624. You can also read the [README in this PR](https://github.com/tgalopin/ux/blob/31e064fb2ab6801fd6bd62b9a1505c6125fb67da/src/Turbo/README.md) to better understand the usage in a Symfony application.

TLDR: Turbo is a generic way to build reactive applications (inline form edition, DOM changes without reload, real-time display of chat messages, etc.) without the need to develop any JavaScript (or at least as little as possible). This has two big advantages: it reduces dramatically the time necessary to build a reactive app and it eliminates in many places the need for a client-side state which could be a source of issues.

This PR is loosely based on the equivalent gem in Ruby that integrates Turbo in RoR. It has however a few key differences:

* it leverages Symfony Mercure by default for Streams instead of Websockets (as PHP isn't a long running process)
* it build upon other tools in Symfony like the `stimulus_controller` [helper introduced a few days ago in Webpack Encore](https://github.com/symfony/webpack-encore-bundle/pull/109)

### CODE REVIEW

File: src/Turbo/Twig/FrameTwigExtension.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @final
+ * @experimental
+ */
+class FrameTwigExtension extends AbstractExtension
```
Comment (nicolas-grekas) [inline src/Turbo/Twig/FrameTwigExtension.php:24]: TurboFrameExtension?

File: src/Turbo/Twig/StreamTwigExtension.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @final
+ * @experimental
+ */
+class StreamTwigExtension extends AbstractExtension
```
Comment (nicolas-grekas) [inline src/Turbo/Twig/StreamTwigExtension.php:26]: TurboStreamExtension?

File: src/Turbo/composer.json [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        "exclude-from-classmap": [
+            "/Tests/"
+        ]
+    },
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/config": "^4.4.17|^5.0",
+        "symfony/dependency-injection": "^4.4.17|^5.0",
+        "symfony/http-kernel": "^4.4.17|^5.0",
+        "symfony/webpack-encore-bundle": "^v1.10.0"
```
Comment (nicolas-grekas) [inline src/Turbo/composer.json:33]: `^1.10`

File: src/Turbo/DependencyInjection/TurboExtension.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+use Symfony\UX\Turbo\Streams\MercureStreamAdapter;
+use Symfony\UX\Turbo\Twig\FrameTwigExtension;
+use Symfony\UX\Turbo\Twig\StreamTwigExtension;
+use Symfony\WebpackEncoreBundle\Twig\StimulusTwigExtension;
+use Twig\Environment;
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @internal
```
Comment (nicolas-grekas) [inline src/Turbo/DependencyInjection/TurboExtension.php:27]: I don't think we need all those `@internal` annotations, do we?
What about `@experimental` also? Some have it, not all.

File: src/Turbo/Twig/FrameTwigExtension.php (hunk 1 above)
Comment (tgalopin) [reply src/Turbo/Twig/FrameTwigExtension.php:24]: I followed the same naming strategy as Webpack Encore Bundle (https://github.com/symfony/webpack-encore-bundle/blob/main/src/Twig/StimulusTwigExtension.php). I quite like having the "TwigExtension" suffix to better understand the difference between DI and Twig extensions in IDEs. 

Adding Turbo as a prefix :)

File: src/Turbo/DependencyInjection/TurboExtension.php (hunk 4 above)
Comment (tgalopin) [reply src/Turbo/DependencyInjection/TurboExtension.php:27]: I may be a bit conservative regarding these annotations indeed :) . Other UX packages have these annotations as well, not sure what our policy should be on them.

Comment (wouterj) [reply src/Turbo/DependencyInjection/TurboExtension.php:27]: I'm often adding `@internal` when the class should not be used in both framework and standalone usage. E.g. when someone is using this package outside the Symfony framework, they probably end up needing to initialize `MercureStreamAdapter`. So I wouldn't add `@internal` there.

As for `@experimental`, I'm often not adding it to `@internal` classes (as there is no BC promise in those anyway). Not sure if these are "Symfony policies" or only my personal policy regarding them :)

**Comment by tgalopin** [comment]
@Kocal I'm not familiar with eslint, any idea why it fails on this package but not others? I may be missing required config but I don't see it

**Comment by Kocal** [comment]
Hi @tgalopin, I think it's because of this https://github.com/tgalopin/ux/blob/ux-turbo/package.json#L34

I don't think this glob pattern will match `src/Turbo/Resources/assets/test/adapters/mercure-controller.test.js` since `mercure-controller.test.js` is under `adapters` folder.

Can you try using ` "src/*/Resources/assets/test/**/*.js"` instead?

**Comment by tgalopin** [comment]
Aah I was looking in the package, I didn't think of this file. Let me try, thanks!

File: src/Turbo/Twig/FrameTwigExtension.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Turbo/Twig/FrameTwigExtension.php:24]: On the other end, none how the bridge's extension have Twig in their suffix:
https://github.com/symfony/symfony/tree/5.x/src/Symfony/Bridge/Twig/Extension
I don't think the suffix is useful personally...

**Comment by nicolas-grekas** [comment]
Missing LICENSE file, see #64 :)

File: .github/workflows/test.yaml [hunk 5]
```diff
@@ -86,6 +93,12 @@ jobs:
                   composer config platform.php 7.4.99
                   composer update --prefer-dist --no-interaction --no-ansi --no-progress
                   php vendor/bin/simple-phpunit
+            - name: Turbo
+              run: |
+                  cd src/Turbo
+                  composer config platform.php 7.4.99
```
Comment (stof) [inline .github/workflows/test.yaml:99]: do we still need this or has the support of PHP 8 been solved in the meantime ?

File: src/Turbo/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                    ->info('Enable this section to use Turbo Streams')
+                    ->children()
+                        ->scalarNode('adapter')
+                            ->defaultNull()
+                            ->info('The service to use to create Turbo Streams')
+                        ->end()
+                        ->arrayNode('options')
+                            ->info('Key/value options passed to the adapter to create Turbo Streams')
+                            ->normalizeKeys(false)
+                            ->variablePrototype()
```
Comment (stof) [inline src/Turbo/DependencyInjection/Configuration.php:41]: this is missing a `->end()` to be consistent with the indentation, to reflect which node is ended each time.

File: src/Turbo/README.md [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    {{ form(message_form) }}
+
+    <!--
+    We can also display another link to go back to the list: the original HTML from frame
+    "message_1" in the list will be displayed again.
+    -->
+    <a href="/messages">Back</a>
+
+<!-- End the frame -->
+{{ turbo_frame_end() }}
```
Comment (stof) [inline src/Turbo/README.md:103]: shouldn't this be a tag rather than a couple of `_start` and `_end` functions ?

Comment (tgalopin) [reply src/Turbo/README.md:103]: AFAIK, tags do things and functions display things, thus this notation. I do agree tags would probably look nicer though. Not fixed about this.

Comment (richardhj) [reply src/Turbo/README.md:103]: Tags would help in IDE autocompletion (same as for 'apply' and 'endapply'), so it's less error-prone I guess.

Comment (nicolas-grekas) [reply src/Turbo/README.md:103]: Do we have any examples of tags that output something?
forms use functions for the same reason.

Comment (wouterj) [reply src/Turbo/README.md:103]: One that is very close to this use-case is NelmioSecurityBundle's CSP tags:

```
{% cspscript %}
<script>
    window.api_key = '{{ api_key }}';
</script>
{% endcspscript %}
```
(the modify the `<script>` tag to add the correct CSP digest)

Comment (tgalopin) [reply src/Turbo/README.md:103]: That's a different use case though, as it modifies the content of the tag (a bit like autoescape would). It doesn't display anything by itself.

I think I would be in favor of keeping in functions. It matches the way the rest of Symfony works (esp. forms) and it doesn't have that big of an impact on DX IMO. IDE can usually autocomplete functions as well as tags.

**Comment by tgalopin** [comment]
Fixed reviews

**Review by nicolas-grekas — COMMENTED** [review]
We might need a pack to install mercure-bundle + ux-turbo at the same time, with a working recipe

File: src/Turbo/Twig/TurboStreamExtension.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $options = array_merge($this->options, ['id' => $id]);
+
+        $html = '<div ';
+        $html .= $this->stimulus->renderStimulusController($env, [
+            $this->adapter->getStimulusControllerName() => $this->adapter->createDataValues($options),
+            'symfony--ux-turbo--core' => [],
+        ]).' ';
+
+        foreach ($attrs as $name => $value) {
```
Comment (nicolas-grekas) [inline src/Turbo/Twig/TurboStreamExtension.php:67]: maybe we don't need attrs at all here btw?

File: src/Turbo/Twig/TurboStreamExtension.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        return [
+            new TwigFunction('turbo_stream_from', [$this, 'streamFrom'], ['needs_environment' => true, 'is_safe' => ['html']]),
+        ];
+    }
+
+    public function streamFrom(Environment $env, string $id, array $attrs = []): string
+    {
+        $options = array_merge($this->options, ['id' => $id]);
+
+        $html = '<div ';
```
Comment (nicolas-grekas) [inline src/Turbo/Twig/TurboStreamExtension.php:61]: I'm wondering: should this generate an html tag, or a list of attributes?

File: src/Turbo/README.md [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+}
+```
+
+```twig
+{# chat.html.twig #}
+{% extends 'base.html.twig' %}
+
+{% block body %}
+    {# This line configures Turbo to listen to the "chat" topic on Mercure #}
+    {{ turbo_stream_from('chat') }}
```
Comment (sroze) [inline src/Turbo/README.md:178]: Though I appreciate you technically want it to live within a `block` in here, semantically it would make greater sense to register it within your Twig file like that:

```twig
{% listen_turbostream "chat" %}
```

**Review by sroze — COMMENTED** [review]
That looks fun!

File: src/Turbo/README.md (hunk 7 above)
Comment (stof) [reply src/Turbo/README.md:103]: Well, these 2 functions are about rendering some HTML opening and closing tags. they are indeed not a kind of twig block

Comment (fbourigault) [reply src/Turbo/README.md:103]: What happens if I forget some `turbo_frame_end()` function calls?

Comment (stof) [reply src/Turbo/README.md:103]: the HTML tag is not closed, and the browser may auto-close it in a different place than where you expect it to be closed.

**Comment by beberlei** [comment]
I think this should be named `symfony/ux-turbo-streams`, because it has nothing to do with Turbo support in general, which works without this PR by just installing the npm/yarn package.

**Comment by tgalopin** [comment]
Closing in favor of the other PR
